django/test/runner.py

```diff
@@ -41,6 +50,119 @@
 except ImportError:
     tblib = None
 
+def recursive_injection(callstack):
+  body2 = []
+
+  for b in callstack.body:
+    #print(b, end="Next, \n")
+    bodyExists = True
+    try:
+      b.body
+    except:
+      bodyExists = False
+
+    if bodyExists:
+      recursive_injection(b)
+    
+    
+    inject_func = ast.parse(f"state_output(locals(), \"{b}\")")
+    body2.append(inject_func)
+    body2.append(b)
+
+    #print(body2)
+  
+  body2.append(inject_func)
+  callstack.body = body2
+
+
+  callStack = True
+  try:
+    callstack.orelse
+  except:
+    callStack = False
+
+  if callStack:
+    orelse2 = []
+
+    for b in callstack.orelse:
+      #print(b, end="Next, \n")
+
+      bodyExists = True
+      try:
+        b.body
+      except:
+        bodyExists = False
+
+      if bodyExists:
+        #print("Sub-body!")
+        recursive_injection(b)
+
+
+      inject_func = ast.parse(f"state_output(locals(), \"{ast.unparse(b)}\")")
```

Comment:
This line is duplicated a few places and should be encapsulated.

In general, if you ever find yourself copying and pasting code, that's a good indication it should be a function.

https://youtu.be/DM2hEBwEWPc?si=3s05CI96g1FDwB7U&t=2458 is interesting. It's about a different programming language but the talk is still interesting.